Java/Dynamic-Programming/DP.java

```diff
@@ -0,0 +1,183 @@
+ find the optimal solution for the next state.
+ 
+ -------------------------------------What does a "state" stand for?---------------------------------
+ "state" is a solution/quantity for a sub problem, it is uniquely identified by parameters. 
+ In this case minimum number of coins which sum up to i, 
+ we can easily find the next state – the solution for i+1. ..... that finally leads to S sum.
+ -------
+ Let's follow Tabulation method(you can take memoizatin - it's your wish)
+ Steps to solve a DP
+1) Identify if it is a DP problem
+
+   - It has overlapping sub problems as we said above
+   - Optimal Substructure: A given problems has Optimal Substructure Property if optimal solution 
+     of the given problem can be obtained by using optimal solutions of its subproblems.
+	 
+2) Decide a state expression with 
+   least parameters
+      
+	- Lets say that n coins are v[j] 0<=j<=n-1. sub-problems sum 1<=i<=S (S is the required sum)   
+	- let Min[sum] stores the min. coins for given sum.
+	
+3) Formulate state relationship    
+     let's say the coins we have are 1,2,4
+	 
+	 for sum 0 - 0 coins -  From this let's consider (Min[0] = 0) as base value. 
+	 psum is present sum.
+	 for sum 1 - 1 coin(v[0])  Min[1] = 1 (oh! lets take it as Min[i]=i)-- we prove it as wrong
+     we can find that coins<=present required sum are to be considered, so we add a condition (v[j]<=i).
+	 for sum 2 - take  v[0](1)<=2  so we proceed, next ?
+	 we get an idea that Min[(psum)i-v[j]] -> Min[2-1] gives num of 1 coins but Min[1] equals 1 instead of 2,
+	 we change it as Min[i] = Min[i-v[j]] + 1
+	 we see that only coin 1 is less than or equal to the current sum. Analyzing it, we see that for sum 1-V[0]= 0 
+	 we have a solution with 0 coins. Because we add one coin to this solution, we’ll have a solution with 1 coin for sum 1
+     Next we get Min[2] = Min[2-2]+1 --->1 coin (i.e; coin 2)
+	 
+	 Hurrah! you can see that it follows optimal substructure property.
+	 
+4) Do tabulation (or add memoization)
+     Let's look at the code! */
+	class Coins
+	{ 
+	 private int cvalue[];
+	  public Coins()
+	  {
+		  
+	  }
+	  public Coins(int n)
+	  {   //let n = 3
+		  cvalue[] = new cvalue[n]; 
+		  cvalue[0] = 1;
+		  cvalue[1] = 2;
+		  cvalue[2] = 4;
+	  }
+	  /* @return minimum number of coins for a given sum */
+	  
+	   public int minCoins(int sum)
+	   {
+		 int Min[] = new int[sum+1]; 
+		                               
+		 for(int i=1;i<=sum;i++)
+			 Min[i] = -1;
+		 
+		 Min[0] = 0;
+		 
+		 for(int i=1;i<=sum;i++)
+		 {
+			for(int j=0;j<cvalue.length;j++)
+			{   
+		       if(cvalue[j]<=i&&Min[i-cvalue[j]]+1){
+				   
+				Min[i] = Min[i-cvalue[j]]+1;
+				
+			    }
+                return Min[sum];				
+			}				
+		 }
+	   }
+	   public static void main(String []args)
+	   {
+		   Coins c = new Coins(3);
+		   System.out.println("Minimum number of coins for the given sum 4: "+c.minCoins(4));
+	   }
+	}
```

Comment:
Please have an EOF

Reply:
Thanks! it is a good practice to have EOF. I have added it